src/targets/maven.ts

```diff
@@ -0,0 +1,352 @@
+   */
+  private async uploadDistribution(distDir: string): Promise<void> {
+    const {
+      targetFile,
+      javadocFile,
+      sourcesFile,
+      pomFile,
+    } = this.getFilesForMavenCli(distDir);
+
+    // Maven central is very flaky, so retrying with an exponential delay in
+    // in case it fails.
+    await retrySpawnProcess(this.mavenConfig.mavenCliPath, [
+      'gpg:sign-and-deploy-file',
+      `-Dfile=${targetFile}`,
+      `-Dfiles=${javadocFile},${sourcesFile}`,
+      `-Dclassifiers=javadoc,sources`,
+      `-Dtypes=jar,jar`,
+      `-DpomFile=${pomFile}`,
+      `-DrepositoryId=${this.mavenConfig.mavenRepoId}`,
+      `-Durl=${this.mavenConfig.mavenRepoUrl}`,
+      `--settings ${this.mavenConfig.mavenSettingsPath}`,
+    ]);
+  }
+
+  /**
+   * Retrieves a record of all the required files by Maven CLI to upload
+   * anything.
+   *
+   * @param distDir directory of the distribution.
+   * @returns record of required files.
+   */
+  private getFilesForMavenCli(distDir: string): Record<string, string> {
+    const moduleName = parse(distDir).base;
+    const targetFile = join(distDir, this.getTargetFilename(distDir));
+    const javadocFile = join(distDir, `${moduleName}-javadoc.jar`);
+    const sourcesFile = join(distDir, `${moduleName}-sources.jar`);
+    const pomFile = join(distDir, 'pom-default.xml');
+    return {
+      targetFile,
+      javadocFile,
+      sourcesFile,
+      pomFile,
+    };
+  }
+
+  /**
+   * Retrieves the target file name for the current distribution.
+   *
+   * If the distibution is an Android distribution, the target file is the
+   * file containing "release" in the name and the ".aar" extension.
+   * Typically, the module (directory) name without the version and appending
+   * "-release.aar" at the end.
+   *
+   * If the distribution isn't an Android distribution, the target filename is
+   * the module name appending ".jar" to the end.
+   *
+   * @param distDir directory where distributions are.
+   * @returns the target file name.
+   */
+  private getTargetFilename(distDir: string): string {
+    const moduleName = parse(distDir).base;
+    const isAndroidDistDir = this.config.android.distDirRegex.test(moduleName);
+    if (isAndroidDistDir) {
+      return moduleName.replace(
+        this.config.android.fileReplaceeRegex,
+        this.config.android.fileReplacerStr
+      );
+    }
+    return `${moduleName}.jar`;
+  }
+
+  /**
+   * Finishes the release flow, by closing and releasing to the repository.
+   * Note that this is a required step to make a release, but it doesn't
+   * perform any release on its own. Required files must have previously been
+   * uploaded accordingly.
+   */
+  public async closeAndRelease(): Promise<void> {
+    // Maven central is very flaky, so retrying with an exponential delay in
+    // in case it fails.
+    await retrySpawnProcess(this.mavenConfig.gradleCliPath, [
+      'closeAndReleaseRepository',
+    ]);
+  }
```

Comment:
I really don't like this one-line function to be abstracted out. I'll move this into the main `publish` in a push.